src/mds/mdstypes.h

```diff
@@ -783,6 +789,18 @@ void inode_t<Allocator>::decode(bufferlist::const_iterator &p)
     export_pin = MDS_RANK_NONE;
   }
 
+  if (struct_v >= 16) {
+    decode(export_ephemeral_random_pin, p);
+  } else {
+    export_ephemeral_random_pin = MDS_RANK_NONE;
+  }
+
+  if (struct_v >= 17) {
+    decode(export_ephemeral_distributed_pin, p);
+  } else {
+    export_ephemeral_distributed_pin = MDS_RANK_NONE;
+  }
```

Comment:
My issue with this change is that these values should not be in `inode_t`; these values should not be part of the struct written to directories in the metadata pool. I know we agreed they should be journaled because an MDS failover should not disrupt the pins but that only means it needs to be part of the journaled metadata. When the journal flushes to the metadata pool directory objects, these values are not persisted.

So what needs to happen is that this extra metadata needs to be part of a secondary structure which is included in `EMetaBlob::fullbit`. It'll be persisted in the journal events but not in the `inode_t`.

Reply:
I was partially confused when I wrote this comment. We **do** want to persist the vxattrs for random/distributed pin settings on a directory inode.

**Whether** the directory inode is ephemerally pinned will be stored in `EMetaBlob::fullbit`. Whether it's ephemerally pinned is indicated if it was randomly selected (the random ephemeral pin) or whether it is part of a distribution (the distributed ephemeral pin). Of course, **where** it's pinned is always determined by the consistent hash of its inode number, so how it was ephemerally pinned doesn't matter. So, we wouldn't expect to need to persist the rank it's pinned to in the journal. We can always recalculate it.